### PR TITLE
appsec: preview version

### PR DESCRIPTION
**Pull request by Julio-Guerra**
### What does this PR do?

Bootstrap a first version of the AppSec agent integrated into the trace agent as a reverse proxy.

### Describe how to test your changes

Usual HTTP server tests

### CODE REVIEW

File: pkg/config/config.go
Comment (gbbr) [inline pkg/config/config.go:908]: I think that for the sake of consistency, all of the "set known" and "bind env" stuff should be in this package, as it is for all other agents. Perhaps you can have a `setupAppsec` function in a separate file which can have your team as the owner.

File: pkg/trace/agent/agent.go
Comment (gbbr) [inline pkg/trace/agent/agent.go:75]: A few notes here, I'm adding numbers in case you want to answer each separately:

1. `coreconfig.Datadog` is a global object accessible from any package, why do you need to have it as an argument? I suspect you might've done it for tests, in which case, I'd say to make the public API `appsec.New` and have that call an internal unexported function `new` which takes the config and then use that for tests. No reason to sacrifice "public" API for testability.
2. I think that this code can safely go into the `api.NewHTTPReceiver` function. This seems like the wrong place.

File: pkg/trace/api/endpoints.go
Comment (gbbr) [inline pkg/trace/api/endpoints.go:90]: Why did you do this? Did you mean to be using https://golang.org/pkg/net/http/#StripPrefix instead?

File: pkg/trace/agent/run.go
Comment (gbbr) [inline pkg/trace/agent/run.go:152]: Please remove this.

File: pkg/trace/api/endpoints.go
Comment (Julio-Guerra) [reply pkg/trace/api/endpoints.go:90]: I tried `http.StripPrefix` indeed, but it doesn't work from here: we end up either with 301 Moved Permanently with a request to `/appsec`; or 404 Not Found with a request to `/appsec/`.

Comment (gbbr) [reply pkg/trace/api/endpoints.go:90]: Could it be that you might have been doing something wrong? Check https://play.golang.org/p/PqdIdFPWchG

File: pkg/config/config.go
Comment (Julio-Guerra) [reply pkg/config/config.go]: Fixed: move `pkg/appsec/config.InitConfig()` to `pkg/config.setupAppSec()` in a new `appsec.go` file (added to CODEOWNERS too)

File: pkg/trace/agent/agent.go
Comment (Julio-Guerra) [reply pkg/trace/agent/agent.go]: Done as stated. I'll create the private `new` function if needed when writing the tests.

File: pkg/trace/agent/run.go
Comment (Julio-Guerra) [reply pkg/trace/agent/run.go]: Done - glitch from the past version ^^'

File: pkg/config/appsec.go
Comment (gbbr) [inline pkg/config/appsec.go:6]: The two options below automatically set things as "known", so you can remove these two lines.

File: pkg/trace/api/api.go
Comment (gbbr) [inline pkg/trace/api/api.go:93]: I think the code is self-explanatory enough to not warrant this comment. Can we remove it please?

File: pkg/trace/api/endpoints.go
Comment (Julio-Guerra) [reply pkg/trace/api/endpoints.go]: Now fixed, but only with `/appsec/` instead of `/appsec` so that it gets stripped to `/` and the following AppSec ServeMux can properly match it.

For the record: the problem I had was coming from a mix if StripPrefix when it returns the empty string, and the AppSec ServeMux which wasn't happy about it because it expects "/" and not "".

File: pkg/config/appsec.go
Comment (Julio-Guerra) [reply pkg/config/appsec.go]: Fixed 👍

File: pkg/trace/api/api.go
Comment (Julio-Guerra) [reply pkg/trace/api/api.go]: Removed 👍

File: pkg/trace/api/api.go
Comment (gbbr) [inline pkg/trace/api/api.go:31]: This import is in the wrong group of imports.

File: pkg/trace/api/api.go
Comment (gbbr) [inline pkg/trace/api/api.go:81]: Can you move this field into the bigger group above please? The one with `out`, `conf`, etc.

File: pkg/trace/api/api.go
Comment (gbbr) [inline pkg/trace/api/api.go:109]: Same here.

File: pkg/appsec/config/config.go
Comment (gbbr) [inline pkg/appsec/config/config.go:11]: I recommend removing this. There are already mechanisms in place to facilitate testing the global configuration that you can use. No need to add extra layers. Furthermore, this isn't used now.

File: pkg/appsec/config/config.go
Comment (gbbr) [inline pkg/appsec/config/config.go:22]: I'd recommend documenting only _what it is_ not _where it is used_.

File: pkg/appsec/api/http/mux.go
Comment (gbbr) [inline pkg/appsec/api/http/mux.go:14]: This is overkill. You have a proxy inside a mux, inside another mux. There are no routes, there is just one proxy. Can we just drop everything and leave only the proxy? In `pkg/trace/endpoints.go` you can then link it straight to `/appsec/proxy`.

**Review by gbbr — CHANGES_REQUESTED** [review]
* As far as `pkg/trace` goes, this is mostly fine, except some small nits.
* As for `pkg/appsec`, I think the current implementation is over engineered. All you need is one simple proxy handler and one single package. There is no need for multiple packages nor setting the ground for speculative future implementations. 

Can we please implement strictly what we need and not add code which is not needed today? That is: one handler and one package, which is a proxy, and that's it. What do you think?

**Review by gbbr — CHANGES_REQUESTED** [review]
P.S. 

* You're going to also need to add metrics (probably best scoped as `datadog.trace_agent.appsec.*` which count incoming requests and their size. 
* You should also time the operation of forwarding the request using `timing.Since`. These will be of very high value for support tickets. 
* We also need to limit requests (you can start with 10MB which I assume should be plenty for a bunch of events - clients can comply).

You will have to add more customisation to the proxy to be able to achieve these...

File: pkg/trace/api/api.go
Comment (Julio-Guerra) [reply pkg/trace/api/api.go]: Fixed

File: pkg/trace/api/api.go
Comment (Julio-Guerra) [reply pkg/trace/api/api.go]: Fixed

File: pkg/appsec/config/config.go
Comment (Julio-Guerra) [reply pkg/appsec/config/config.go]: Done: replaced by the existing interface type `(pkg/config).Config`

File: pkg/appsec/config/config.go
Comment (Julio-Guerra) [reply pkg/appsec/config/config.go]: I think it does with "intake site applied by default to [the default intake url template] defaultIntakeURLTemplate.
I tried to make it more verbose. Let me know if anything else.

File: pkg/appsec/api/http/mux.go
Comment (Julio-Guerra) [reply pkg/appsec/api/http/mux.go]: Done: now simplified to a single`appsec` package.

File: pkg/trace/api/api.go
Comment (Julio-Guerra) [reply pkg/trace/api/api.go]: Fixed

File: pkg/trace/api/api.go
Comment (gbbr) [inline pkg/trace/api/api.go:94]: All log messages need to be capitalised.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:47]:
```suggestion
	via := fmt.Sprintf("trace-agent %s", info.Version)
```

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:36]: I don't think this qualifies as an error messages. Disabling appsec is a potentially valid user intention.
```suggestion
		return disabled, nil
```

File: pkg/trace/api/api.go
Comment (gbbr) [inline pkg/trace/api/api.go:92]:
```suggestion
	appsecHandler, err := appsec.NewIntakeReverseProxy(conf.NewHTTPTransport())
```
👍🏻

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:64]:
```suggestion
		AppSecRequestMetricsPrefix = "datadog.trace_agent.appsec."
```
We already have too many namespaces as is. No need to go that deep. We can then follow with:
* `.request`
* `.request_duration_ms`
* `.request_bytes`

This will match other metrics in convention

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:67]: What's the point of having these constants? Can't we just inline them? They are only used once...

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:79]: Nice 👍🏻

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:84]: Please use `timing.Since`

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:59]: This is a nice mechanism, but it fails if `Content-Length` is not set, and is inaccurate if for example the reader limit is reached. I'm afraid you won't be able to get away without writing the actual transport, as we do for profiles. Another requirement which asks for it is that with this approach, we're not able to create metrics for errors/failures.

**Review by gbbr — COMMENTED** [review]
Thanks Julio. This is much better in my opinion.

File: pkg/appsec/config/config.go
Comment (gbbr) [reply pkg/appsec/config/config.go]: What I meant is:
```
// defaultSite is the default site used to target the intake.
```
Or similar. No need to mention the other constant, and how it is used. As far as we know, it may be used anywhere and it's not restricted to that purpose. But do as you please, I don't feel strongly here.

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go:69]: Now that we rely on `LimitedReader`, this code was replaced with an access to its `Count` field.
As of the reached limit error, the `ErrorHandler` field of the reverse proxy does the job and allows to know when an error occurred during the `io.Copy` it does (tests will cover this case).

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go:91]: I'd love to, but I don't see how to pass it the tags.

File: pkg/appsec/config/config.go
Comment (Julio-Guerra) [reply pkg/appsec/config/config.go]: I see. Fixed.

File: pkg/trace/api/api.go
Comment (Julio-Guerra) [reply pkg/trace/api/api.go:95]: Fixed

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Fixed.

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Ok, but I also added a log message before returning.

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Fixed.

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Now replaced by the actual body read size taken from the Count field of the LimitedReader struct

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:19]: This import needs to be in its own group (external import). Just needs an empty line above it.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:22]: This is no longer used.

**Review by gbbr — COMMENTED** [review]
Cool. I think we can move on to adding tests. Besides unit tests, you might find it valuable to potentially extend the `pkg/trace/test` package to add a fake backend intake for appsec and ensure black box E2E testing. 

As for unit tests, please make sure to cover all ground, including all error branches, returned responses and generated metrics (see `(pkg/trace/test/testutil).TestStatsClient`)

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Just willing to avoid copy/paste mistakes + grouping all of them together to make it clear + relying on the compiler to do the concatenation at compilation time

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Fixed

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Indeed. Fixed.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:29]: I think you can just use `http.Error` here.

File: pkg/trace/api/endpoints.go
Comment (nizox) [inline pkg/trace/api/endpoints.go:84]: Can we use `/appsec/v1/input` instead?

In fact, as the agent is only a proxy for now, the `v1/input` suffix defines the "event sending protocol" while the base URL `http://localhost:8126/appsec/` & `https://appsecevts-http-intake.logs.datad0g.com/` are destinations.

The libraries could then send events to the agent or intake directly only by changing the base URL, which is useful for development and makes implementation more straightforward.

Comment (Julio-Guerra) [reply pkg/trace/api/endpoints.go:84]: I'd rather have `/appsec/v0.1/proxy` forwarding to `https://appsecevts-http-intake.logs.datad0g.com` so that the agent API endpoint to use becomes `/appsec/v0.1/proxy/v1/input`

wdyt @gbbr? the base problem comes from the case when a library is in agentless mode. This would simplify a bit the configuration by just having to change the API URL "prefix" part (eg. http://localhost:8126/appsec/v0.1/proxy/v1/input would become https://appsecevts-http-intake.logs.datad0g.com/v1/input)

Comment (gbbr) [reply pkg/trace/api/endpoints.go:84]: I think `/appsec/v1/input` is the right way to go. We use the same for profiling: `/profiling/v1/input`.

Comment (Julio-Guerra) [reply pkg/trace/api/endpoints.go:84]: by the way, this is why the profiling api endpoint of the trace-agent is `/profiling/v1/input`: to have the same endpoint URL suffix as the intake's

Comment (gbbr) [reply pkg/trace/api/endpoints.go:84]: I'd also be ok with `/appsec/proxy/v1/input`. That would still hold for the example given by @nizox. I think having a double version in one single URL is a bit much.

Comment (Julio-Guerra) [reply pkg/trace/api/endpoints.go]: Done: `/appsec/v0.1/proxy` becomes `/appsec/v1/input`
The v0.1 Agent API versioning is dismissed for now. We'll reintroduce it in the future if needed by an actual agent API.

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go:29]: The intake api spec tells it returns a json value, even in case of error (explaining why), so I was thinking in aligning here.

File: pkg/trace/api/endpoints.go
Comment (Julio-Guerra) [reply pkg/trace/api/endpoints.go]: I think we are good to go with this during for the preview version.
There won't be any extra appsec endpoint as far as anyone could tell me.

**Review by albertvaka — APPROVED** [review]
👍 changes to codeowners

**Review by hush-hush — APPROVED** [review]

**Review by nizox — APPROVED** [review]
Tested E2E, all good.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:90]: We should count metrics for _all_ errors, not just for this one. A tag should specify which kind of error it is.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:91]: Frankly we never check the error return value for this, anywhere 🙂

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:124]: The cardinality for this tag is too high. You'll have to remove it. It's already counted as a metric anyway.

File: pkg/trace/api/api.go
Comment (gbbr) [reply pkg/trace/api/api.go:95]: Not 🙂 

Also, there is no "appsec agent". Rather say:
```suggestion
		log.Errorf("Could not instantiate AppSec: %v", err)
```

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:89]: You could just do this above, where you define the ErrorHandler, but I'll leave it up to you...

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go:91]: Ok, so we need to adjust errcheck in .golangci file tell not to fail on these calls, like it's done for `log.Error` =)

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: I did a change to monitor every error.
But getting the type of the error to distinguish them is sometimes too weak, for example when the error is created with `errors.New()` and the type doesn't describe the error.
Cf.  https://github.com/DataDog/datadog-agent/pull/8359/files#diff-204368e0d1216a43867fe43ec306add180e68d2b9db3414c95c0c08fd4d46ff0R90-R95

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: I changed `.golangci` so that `errcheck` doesn't trigger for calls to functions from `pkg/trace/metrics`.
And I simplified back the code by removing the error checks.

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Fixed

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go:88]: Yes, just for the sake of having all the metrics in `withMetrics` :-)

Comment (gbbr) [reply pkg/appsec/appsec.go:88]: Seems like a lot of hassle for little gain. If you would re-write the proxy transport you could have it all in one place. I don't feel strongly about it, as long as it works!

Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: You're right, a custom RoundTripper is indeed clearer and cleaner 👍
Now fixed.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:115]: This scares me a little. A race condition (panic) is possible here. See: https://github.com/golang/go/blob/go1.16/src/net/http/client.go#L134-L138.

We have 2 options:
* Alter the `LimitedReader` struct to access `Count` under a lock (and then retrieve it using a method `Count()`)
* Give up on this metric entirely

**Review by gbbr — COMMENTED** [review]
Almost there, one last concern :)

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Removed as we couldn't find a proper way of making LimitedReader atomic.

File: pkg/appsec/appsec.go
Comment (gbbr) [inline pkg/appsec/appsec.go:39]: This sounds like we're telling the user that something is wrong and that they should take action. I think the message: "AppSec proxy disabled." should be enough information. They already know the settings since they clearly manually set it to `false` (being it enabled by default).

File: pkg/appsec/config.go
Comment (gbbr) [inline pkg/appsec/config.go:42]: Is the value of `0` acceptable? If not, what should it mean? Should it mean "unlimited"? Probably not a good idea. I'd say let's set this only if its `> 0`. WDYT?

Comment (Julio-Guerra) [reply pkg/appsec/config.go]: Yes, it was unlimited if <= 0
I changed it with a default payload size of 5 MB when <= 0

File: pkg/appsec/appsec.go
Comment (Julio-Guerra) [reply pkg/appsec/appsec.go]: Good point - fixed

**Review by gbbr — APPROVED** [review]
🎉 Awesome! Thanks Julio!

P.S. No merging yet please, we are in code freeze 🥶

File: pkg/config/config_template.yaml
Comment (ruthnaebeck) [inline pkg/config/config_template.yaml:908]:
```suggestion
## Uncomment the parameters `appsec_config` and `enabled` to enable AppSec.
```

File: pkg/config/config_template.yaml
Comment (ruthnaebeck) [inline pkg/config/config_template.yaml:913]:
```suggestion
  ## Enable the AppSec proxy in the APM Agent.
```

File: pkg/config/config_template.yaml
Comment (ruthnaebeck) [inline pkg/config/config_template.yaml:918]:
```suggestion
  ## Define the endpoint and port to hit when using a proxy for AppSec. Logs are forwarded by TCP;
```

File: pkg/config/config_template.yaml
Comment (ruthnaebeck) [inline pkg/config/config_template.yaml:919]:
```suggestion
  ## therefore, the proxy must be able to handle TCP connections.
```

File: pkg/config/config_template.yaml
Comment (ruthnaebeck) [inline pkg/config/config_template.yaml:924]:
```suggestion
  ## Define the maximum HTTP payload size (the body) in bytes the AppSec proxy allows.
```

File: releasenotes/notes/appsec-proxy-8e34c687cfc23e51.yaml
Comment (ruthnaebeck) [inline releasenotes/notes/appsec-proxy-8e34c687cfc23e51.yaml:1]:
```suggestion
# Each section from every release note is combined when the
```

**Review by ruthnaebeck — CHANGES_REQUESTED** [review]
Docs review

**Review by ruthnaebeck — APPROVED** [review]
👍  for docs

**Pull request merged**
